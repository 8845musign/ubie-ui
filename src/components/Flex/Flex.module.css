.flex {
  --gap: 0;
  --align-items: center;
  --justify-content: flex-start;

  display: flex;
  gap: var(--gap);
  align-items: var(--align-items);
  justify-content: var(--justify-content);
}
