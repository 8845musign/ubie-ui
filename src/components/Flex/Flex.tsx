'use client';

import clsx from 'clsx';
import { isValidElement, cloneElement } from 'react';
import styles from './Flex.module.css';
import { CustomDataAttributeProps } from '../../types/attributes'; // 追加したインポート
import { Spacing, AlignItems, JustifyContent, FlexDirection } from '../../types/style';
import { paddingVariables, marginVariables, gapVariables } from '../../utils/style';
import { HTMLTagname } from '../../utils/types';
import { Box } from '../Box/Box';
import type { PaddingProps, MarginProps } from '../../types/style';
import type { FC, PropsWithChildren, ReactElement, ComponentType, ReactNode } from 'react';

type Props = {
  /**
   * レンダリングされるHTML要素
   * @default div
   */
  as?: HTMLTagname | ReactElement<ComponentType<typeof Box>>;
  /**
   * 子要素同士の間隔。指定しない場合は0
   * xxs=4px, xs=8px, sm=12px, md=16px, lg=24px, xl=40px, xxl=64px
   */
  spacing?: Spacing;
  /**
   * direction 重ねる向き
   * @default row
   */
  direction?: FlexDirection;
  /**
   * 主軸方向における子要素のレイアウトを定める。`direction` prop が `column` の場合は水平軸、 `row` の場合は垂直軸のレイアウトを制御する。水平軸の場合に、ブロックレベル要素を幅いっぱいに占有させたい場合は `normal` を使うこと
   * @default flex-start
   */
  alignItems?: AlignItems;
  /**
   * 交差軸方向における子要素のレイアウトを定める。directionが `column` の場合は垂軸直、`row` の場合は水平軸のレイアウトを制御する。水平軸の場合に、ブロックレベル要素を幅いっぱいに占有させたい場合は `normal` を使うこと
   * @default flex-start
   */
  justifyContent?: JustifyContent;
  /**
   * 折り返しの有無
   * @default false
   */
  wrap?: boolean;
  /**
   * 親要素に対し、100%としたい場合に使用
   */
  height?: 'full';
  /**
   * デフォルト<Flex>は横幅いっぱいを専有する。しかし例えば、フレックスコンテナの子要素として配置した場合、横幅が自身の子に合わせて小さくなる。これが不都合の場合に100%とする事が可能
   */
  width?: 'full';
} & MarginProps &
  PaddingProps &
  CustomDataAttributeProps;

export const Flex: FC<PropsWithChildren<Props>> = ({
  as: FlexCopmonent = 'div',
  children,
  direction = 'row',
  alignItems = 'flex-start',
  justifyContent = 'flex-start',
  wrap,
  spacing,
  height,
  width,
  p,
  px,
  py,
  pt,
  pr,
  pb,
  pl,
  m,
  mx,
  my,
  mt,
  mr,
  mb,
  ml,
  ...otherProps
}) => {
  /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
  const createElement = (props: any, children: ReactNode) => {
    if (isValidElement(FlexCopmonent)) {
      return cloneElement(FlexCopmonent, FlexCopmonent.props, <div {...props}>{children}</div>);
    } else {
      return <FlexCopmonent {...props}>{children}</FlexCopmonent>;
    }
  };

  return createElement(
    {
      className: clsx(styles.flex, height === 'full' && styles.heightFull, width === 'full' && styles.widthFull),
      style: {
        '--flex-direction': direction,
        '--align-items': alignItems,
        '--justify-content': justifyContent,
        '--flex-wrap': wrap ? 'wrap' : 'nowrap',
        ...gapVariables(spacing),
        ...paddingVariables({
          p,
          px,
          py,
          pt,
          pr,
          pb,
          pl,
        }),
        ...marginVariables({
          m,
          mx,
          my,
          mt,
          mr,
          mb,
          ml,
        }),
      },
      ...otherProps,
    },
    children,
  );
};
