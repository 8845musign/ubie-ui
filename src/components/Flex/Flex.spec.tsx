import { render, screen } from '@testing-library/react';
import { Flex } from './Flex';

describe('<Flex>', () => {
  it('has all the margins through m prop', () => {
    render(
      <Flex m="xxs" data-testid="box">
        test
      </Flex>,
    );
    const div = screen.getByTestId('box');

    expect(div).toHaveStyle('--margin-top: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--margin-right: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--margin-bottom: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--margin-left: var(--size-spacing-xxs)');
  });

  it('has a horizontal margins through mx prop', () => {
    render(
      <Flex mx="xxs" data-testid="box">
        Test
      </Flex>,
    );
    const div = screen.getByTestId('box');

    expect(div).toHaveStyle('--margin-top: 0');
    expect(div).toHaveStyle('--margin-right: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--margin-bottom: 0');
    expect(div).toHaveStyle('--margin-left: var(--size-spacing-xxs)');
  });

  it('has a vertical margins of through mx prop', () => {
    render(
      <Flex my="xxs" data-testid="box">
        Test
      </Flex>,
    );
    const div = screen.getByTestId('box');

    expect(div).toHaveStyle('--margin-top: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--margin-right: 0');
    expect(div).toHaveStyle('--margin-bottom: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--margin-left: 0');
  });

  it('has all margins through individual margin props', () => {
    render(
      <Flex mt="xxs" mr="xs" mb="sm" ml="md" data-testid="box">
        Test
      </Flex>,
    );
    const div = screen.getByTestId('box');

    expect(div).toHaveStyle('--margin-top: var(--size-spacing-xxs)');
    expect(div).toHaveStyle('--margin-right: var(--size-spacing-xs)');
    expect(div).toHaveStyle('--margin-bottom: var(--size-spacing-sm)');
    expect(div).toHaveStyle('--margin-left: var(--size-spacing-md)');
  });

  it('gives priority to individual margin props', () => {
    render(
      <Flex m="xxs" mx="xs" my="xs" mt="sm" mr="md" mb="lg" ml="xl" data-testid="box">
        Test
      </Flex>,
    );
    const div = screen.getByTestId('box');

    expect(div).toHaveStyle('--margin-top: var(--size-spacing-sm)');
    expect(div).toHaveStyle('--margin-right: var(--size-spacing-md)');
    expect(div).toHaveStyle('--margin-bottom: var(--size-spacing-lg)');
    expect(div).toHaveStyle('--margin-left: var(--size-spacing-xl)');
  });
});
