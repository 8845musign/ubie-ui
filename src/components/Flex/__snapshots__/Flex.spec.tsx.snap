// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Flex> > Snapshot Test > changes rendered elements 1`] = `
<section
  class="_flex_fd9099"
  data-testid="flex"
  style="--flex-direction: row; --align-items: flex-start; --justify-content: flex-start; --flex-wrap: nowrap; --gap: 0; --padding-top: 0; --padding-right: 0; --padding-bottom: 0; --padding-left: 0; --margin-top: 0; --margin-right: 0; --margin-bottom: 0; --margin-left: 0; --width: auto; --min-width: auto; --max-width: 100px;"
>
  <div
    class="_flexItem_c2f191 _longhand_c2f191"
    style="--min-width: auto; --max-width: none; --flex-grow: 1; --flex-shrink: 1; --flex-basis: auto; --padding-top: 0; --padding-right: 0; --padding-bottom: 0; --padding-left: 0; --margin-top: 0; --margin-right: 0; --margin-bottom: 0; --margin-left: 0;"
  >
    test
  </div>
  <div
    class="_flexItem_c2f191 _longhand_c2f191"
    style="--min-width: auto; --max-width: none; --flex-grow: 1; --flex-shrink: 1; --flex-basis: auto; --padding-top: 0; --padding-right: 0; --padding-bottom: 0; --padding-left: 0; --margin-top: 0; --margin-right: 0; --margin-bottom: 0; --margin-left: 0;"
  >
    test
  </div>
</section>
`;

exports[`<Flex> > Snapshot Test > changes the rendered component 1`] = `
<div
  class="_flex_fd9099"
  data-testid="flex"
  style="--flex-direction: row; --align-items: flex-start; --justify-content: flex-start; --flex-wrap: nowrap; --gap: 0; --padding-top: 0; --padding-right: 0; --padding-bottom: 0; --padding-left: 0; --margin-top: 0; --margin-right: 0; --margin-bottom: 0; --margin-left: 0; --width: auto; --min-width: auto; --max-width: 100px;"
>
  <div
    class="_flexItem_c2f191 _longhand_c2f191"
    style="--min-width: auto; --max-width: none; --flex-grow: 1; --flex-shrink: 1; --flex-basis: auto; --padding-top: 0; --padding-right: 0; --padding-bottom: 0; --padding-left: 0; --margin-top: 0; --margin-right: 0; --margin-bottom: 0; --margin-left: 0;"
  >
    test
  </div>
  <div
    class="_flexItem_c2f191 _longhand_c2f191"
    style="--min-width: auto; --max-width: none; --flex-grow: 1; --flex-shrink: 1; --flex-basis: auto; --padding-top: 0; --padding-right: 0; --padding-bottom: 0; --padding-left: 0; --margin-top: 0; --margin-right: 0; --margin-bottom: 0; --margin-left: 0;"
  >
    test
  </div>
</div>
`;
